File: src/tp3.py
#coding: utf-8
# Implementación en Python del TP3 para prototipos y demas yerbas

import itertools
import sys

from numpy import matrix, sign, zeros, empty
from numpy.linalg import norm, solve
from scipy.sparse import lil_matrix


REMAIN_FACTOR = 0.95
TELEPORTATION_FACTOR = 1 - REMAIN_FACTOR


def load_data(raw_data):
    """
    Carga una matrix con la estructura de una web dada

    El formato de archivo es:

        <n cantidad de paginas>
        <m cantidad de links>
        <l0 from> <l0 to>  link de la pagina l0.from a l0.to
        ...
        <lm from> <lm to>  link de la pagina lm.from a lm.to

    Devuelve una matriz esparsa donde la columna i indica la probabilidad de
    que un surfer en la pagina i vaya a cada pagina de la columna.

    M[i, j] = probabilidad de estando en la pagina j ir a la pagina i.

    Notar que estamos creando la matriz A de Kamvar2003 o sea, es transpuesta.
    """
    pages = int(raw_data.readline())
    links = int(raw_data.readline())

    print 'Armando la matriz esparsa...'
    graph = lil_matrix((pages, pages), dtype='d')

    row_outdegrees = zeros(pages)
    count = 0
    for i in range(links):
        # leo cada linea del archivo
        # corto en el espacio entre numeros
        # casteo a int
        # empiezo por 0 en vez de por 1
        src, dest = [int(s) - 1 for s in raw_data.readline().split()]
        graph[dest, src] = 1
        row_outdegrees[src] += 1

        # helper para ver por donde voy
        count += 1
        if count % 10000 == 0:
            print "%s of %s links loaded..." % (count, links)

    print "Normalizing out edges..."
    # get the used rows and cols
    rows, cols = graph.nonzero()
    count = 0
    for i, j in itertools.izip(rows, cols):
        graph[i, j] = 1.0 / row_outdegrees[j]
        # this can be optimized to only perform the div once per row

        # helper para ver por donde voy
        count += 1
        if count % 10000 == 0:
            print "%s of %s links normalized..." % (count, links)

    print "Matrix loaded. converting to CSR..."
    # we convert it to CSR for faster arithmetic
    graph = graph.tocsr()

    return graph, pages, links, row_outdegrees


def v(n):
    v = empty(n)
    v.fill(1.0 / n)
    return v


def D(web_outdegrees):
    n = len(web_outdegrees)
    d = matrix([[0 if row != 0 else 1 for row in web_outdegrees]]).T
    return d * v(n).T


def P_1(P, web_outdegrees):
    return P + D(web_outdegrees)


def P_2(P_1, v):
    e = matrix([[1] * P_1.shape[0]])
    E = e.T * v.T

    c = REMAIN_FACTOR
    P_2 = c * P_1 + (1 - c) * E
    return P_2


def pagerank_power_kamvar(P, x, criteria, epsilon, remain_factor):
    """
    Perform the power method for the PageRank matrix
    using Kamvar's optimization for matrix multiplication.

    P is the non-stochastic PageRank P matrix
    outdegrees is the vector with the count of outdegrees per
    x is the initial guess for the solution
    flow_factor 1 - TELEPORTATION_FACTOR
    """
    MAX_ITERS = 10000

    k = 1
    delta = 1000000.0  # At least one iteration will be performed

    n = len(x)
    uniform_prob = 1.0 / n

    results = [x]
    while delta >= epsilon and k < MAX_ITERS:
        # Optimized Ax multiplication from algorithm 1
        y = P.dot(remain_factor * x)
        w = norm(x, 1) - norm(y, 1)
        x_k = y + w * uniform_prob

        if criteria == 'abs':
            delta = norm(x_k - x)
        else:
            delta = norm(x_k - x) / norm(x)

        k += 1
        x = x_k
        results.append(x_k)
    for result in results[:-1]:
        print norm(abs(x_k - result), 1)

    return x_k


def solve_gammas(Y, y_k):
    return qr_two_iterations(Y, matrix(y_k).T).T.tolist()[0]


def quad_extrapolation(x_3, x_2, x_1, x_k):
    # print "Performing interpolation..."
    y_2 = x_2 - x_3
    y_1 = x_1 - x_3
    y_k = x_k - x_3

    Y = matrix((y_2, y_1)).T
    gamma_3 = 1

    (gamma_1, gamma_2) = solve_gammas(Y, y_k)

    # gamma_0 = -(gamma_1 + gamma_2 + gamma_3)
    beta_0 = gamma_1 + gamma_2 + gamma_3
    beta_1 = gamma_2 + gamma_3
    beta_2 = gamma_3
    x = beta_0 * x_2 + beta_1 * x_1 + beta_2 * x_k
    return x


def power_quad(P, x, criteria, epsilon, quad_freq, remain_factor):
    """
    Perform the power method for the PageRank matrix
    using Kamvar's optimization for matrix multiplication
    and quadratic optimization.

    P is the stochastic PageRank P' matrix
    x is the initial guess for the solution
    flow_factor 1 - TELEPORTATION_FACTOR
    """
    MAX_ITERS = 10000

    k = 1
    delta = 1000000.0  # At least one iteration will be performed

    # vec = v(web.shape[0])  # we can just multiply by the const 1/n
    prob_teleport = 1.0 / P.shape[0]

    x_3 = x
    x_2 = x
    x_1 = x
    x_k = x

    results = [x]
    while delta >= epsilon and k < MAX_ITERS:
        # Optimized Ax multiplication from algorithm 1
        y = P.dot(remain_factor * x_k)
        w = norm(x_k, 1) - norm(y, 1)
        x_k = y + w * prob_teleport

        if k in quad_freq:
            # print "Performing quad extrapolation..."
            x_k = quad_extrapolation(x_3, x_2, x_1, x_k)

        if criteria == 'abs':
            delta = norm(x_k - x_1)
        else:
            delta = norm(x_k - x_1) / norm(x_1)

        x_3 = x_2
        x_2 = x_1
        x_1 = x_k
        # print "Iter -> %s %s %s <- delta / do quad" % (k, delta, do_quad)
        k += 1
        results.append(x_k)
    for result in results[:-1]:
        print norm(abs(x_k / norm(x_k, 1) - result / norm(result, 1)), 1)
    return x_k


# Just two iterations of it
def qr_two_iterations(A, b):
    from ipdb import set_trace; set_trace();
    m, n = A.shape

    # First iteration
    x = A[:, 0]

    v = zeros((m, 1))
    v[0] = 1.0

    alpha = sign(x[0, 0]) * norm(x, 2)

    v = x - (alpha * v)

    v = v / norm(v, 2)

    A = A - (2 * v * (v.T * A))
    b = b - (2 * v * (v.T * b))

    # Second iteration
    x = A[1:m, 1:n][:, 0]

    v = zeros((m - 1, 1))
    v[0] = 1.0

    alpha = sign(x[0, 0]) * norm(x, 2)

    v = x - (alpha * v)

    v = v / norm(v, 2)

    A[1:m, 1:n] = A[1:m, 1:n] - (2 * v * (v.T * A[1:m, 1:n]))
    b[1:m] = b[1:m] - (2 * v * (v.T * b[1:m]))

    return solve(A[0:2, 0:2], -b[0:2])

if __name__ == '__main__':
    filename = sys.argv[1]
    raw_data = open(filename, 'r')

    print "Loading data..."
    web, pages, links, out_degrees = load_data(raw_data)

    print "Computing PageRank with regular Power Method... c=.9"
    res = pagerank_power_kamvar(web, v(pages), 'rel', 0.0001, 0.9)
    print 'result:\n', res

    print "Computing PageRank with regular Power Quad...c=.9, solo una vez en la 5ta iter"
    res = power_quad(web, v(pages), 'rel', 0.0001, [5], 0.9)
    print res / norm(res, 1)  # es necesario normalizar?

    print
    print "Computing PageRank with regular Power Quad...c=.9, cada 10 iter"
    res = power_quad(web, v(pages), 'rel', 0.0001, range(10, 1000, 10), 0.9)
    print res / norm(res, 1)  # es necesario normalizar?

    print
    print "Computing PageRank with regular Power Quad...c=.9, cada 5 iter"
    res = power_quad(web, v(pages), 'rel', 0.0001, range(5, 1000, 5), 0.9)
    print res / norm(res, 1)  # es necesario normalizar?

    print
    print "Computing PageRank with regular Power Quad...c=.9, cada 4 iter"
    res = power_quad(web, v(pages), 'rel', 0.0001, range(4, 1000, 4), 0.9)
    print res / norm(res, 1)  # es necesario normalizar?

    ########################### c=.95
    print
    print "Computing PageRank with regular Power Method... c=.95"
    res = pagerank_power_kamvar(web, v(pages), 'rel', 0.0001, 0.95)
    print 'result:\n', res

    print
    print "Computing PageRank with regular Power Quad...c=.95, solo una vez en la 5ta iter"
    res = power_quad(web, v(pages), 'rel', 0.0001, [5], 0.95)
    print res / norm(res, 1)  # es necesario normalizar?

    print
    print "Computing PageRank with regular Power Quad...c=.95, cada 10 iter"
    res = power_quad(web, v(pages), 'rel', 0.0001, range(10, 1000, 10), 0.95)
    print res / norm(res, 1)  # es necesario normalizar?

    print
    print "Computing PageRank with regular Power Quad...c=.95, cada 5 iter"
    res = power_quad(web, v(pages), 'rel', 0.0001, range(5, 1000, 5), 0.95)
    print res / norm(res, 1)  # es necesario normalizar?

    print
    print "Computing PageRank with regular Power Quad...c=.95, cada 4 iter"
    res = power_quad(web, v(pages), 'rel', 0.0001, range(4, 1000, 4), 0.95)
    print res / norm(res, 1)  # es necesario normalizar?

    ########################### c=.99
    print
    print "Computing PageRank with regular Power Method... c=.99"
    res = pagerank_power_kamvar(web, v(pages), 'rel', 0.0001, 0.99)
    print 'result:\n', res

    print
    print "Computing PageRank with regular Power Quad...c=.99, solo una vez en la 5ta iter"
    res = power_quad(web, v(pages), 'rel', 0.0001, [5], 0.99)
    print res / norm(res, 1)  # es necesario normalizar?

    print
    print "Computing PageRank with regular Power Quad...c=.99, cada 10 iter"
    res = power_quad(web, v(pages), 'rel', 0.0001, range(10, 1000, 10), 0.99)
    print res / norm(res, 1)  # es necesario normalizar?

    print
    print "Computing PageRank with regular Power Quad...c=.99, cada 5 iter"
    res = power_quad(web, v(pages), 'rel', 0.0001, range(5, 1000, 5), 0.99)
    print res / norm(res, 1)  # es necesario normalizar?

    print
    print "Computing PageRank with regular Power Quad...c=.99, cada 4 iter"
    res = power_quad(web, v(pages), 'rel', 0.0001, range(4, 1000, 4), 0.99)
    print res / norm(res, 1)  # es necesario normalizar?

    print
    print "Los siguientes dos casos deberian dar lo mismo y no"
    print
    print "Computing PageRank with regular Power Quad...c=.99, cada 4 iter"
    res = power_quad(web, v(pages), 'rel', 0.0001, [], 0.85)
    print res / norm(res, 1)  # es necesario normalizar?

    print
    print "Computing PageRank with regular Power Quad...c=.99, cada 4 iter"
    res = power_quad(web, v(pages), 'rel', 0.0001, [4], 0.85)
    print res / norm(res, 1)  # es necesario normalizar?


    # Uncomment to test result
    # print P2.T.dot(res) / norm(res, 2)

File: src/tp3.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <cmath>
#include <memory>
#include <vector>
#include <string>
#include <functional>

#include "sparse_matrix.h"

using namespace std;

const int MAX_ITERS = 500;
const double REMAIN_FACTOR = 0.99;

/*
 * Compute Q transposed and R
 * Solve least squares problem
 */
std::pair<double, double> qr_two_iterations(matrix& Y, vec& b) {
    int m = Y.size();
    int n = Y[0].size();

    auto x = make_unique<vec> (vec(m));
    for (int i=0; i < m; i++) {
        (*x)[i] = Y[i][0];
    }

    // V is the first canonical vector
    unique_ptr<vec> v (new vec(m, 0));
    (*v)[0] = 1;

    auto alpha = ((*x)[0] / abs((*x)[0])) * norm((*x), 2);
    mult_inplace(*v, -alpha);
    sum_inplace(*v, *x);
    div_inplace(*v, norm(*v, 2));

    matrix & A = Y;
    vec& r = b;

    unique_ptr<vec> vt_A (left_trans_multiply(*v, A));
    double vt_B = left_trans_multiply(*v, r);

    unique_ptr<matrix> vvtA (mult_transposed(*v, *vt_A));
    unique_ptr<vec> vvtB = make_unique<vec> ((*v) * (2 * vt_B));

    mult_inplace(*vvtA, 2);
    sub_inplace(A, *vvtA);

    unique_ptr<vec> vvtBd = move(vvtB);
    sub_inplace(r, *vvtBd);
    vec& r_res = r;
    unique_ptr<vec> x_2 (new vec(m - 1, 0));

    for (int i=0; i < A.size() - 1; i++) {
        (*x_2)[i] = A[i + 1][1];
    }

    unique_ptr<vec> v_2 (new vec(m - 1, 0));
    (*v_2)[0] = 1;

    auto alpha_2 = ((*x_2)[0] / abs((*x_2)[0])) * norm((*x_2), 2);

    mult_inplace(*v_2, -alpha_2);

    sum_inplace(*v_2, *x_2);

    div_inplace(*v_2, norm(*v_2, 2));

    unique_ptr<matrix> A_2 (submatrix(A));

    // Generate orthogonal stuff
    vec r_res_sub = vec(r_res.begin() + 1, r_res.end());

    unique_ptr<vec> vt_A2 (left_trans_multiply(*v_2, *A_2));

    double vt_B2 = left_trans_multiply(*v_2, r_res_sub);

    unique_ptr<matrix> vvtA2 (mult_transposed(*v_2, *vt_A2));
    vec vvtB2 = vt_B2 * (*v_2);

    mult_inplace(*vvtA2, 2);

    sub_inplace(*A_2, *vvtA2);

    mult_inplace(vvtB2, 2);
    sub_inplace(r_res_sub, vvtB2);
    vec& r_2 = r_res_sub;

    for(int i=1; i < r.size(); i++) {
        r_res[i] = r_2[i-1];
    }

    insert_block(A, *A_2, 1, 1);

    mult_inplace(r_res, -1);
    vec& inverted_r = r_res;

    return solve_square_eq(A, inverted_r);
}

/* Build Y as a combination of two column vectors */
matrix* build_Y(vec const& y_1, vec const& y_2) {
    // Y is row_major
    auto Y = new matrix(y_1.size(), vec(2, 0));
    for (int i=0; i < y_1.size(); i++) {
        (*Y)[i][0] = y_1[i];
        (*Y)[i][1] = y_2[i];
    }

    return Y;
}

unique_ptr<vec> quad_extrapolation(vec const& x_3,
        vec const& x_2, vec const& x_1, vec const& x_k){
    vec y_2 = x_2 - x_3;
    vec y_1 = x_1 - x_3;
    vec y_k = x_k - x_3;

    unique_ptr<matrix> Y (build_Y(y_2, y_1));

    auto gamma_3 = 1.0;
    double gamma_1;
    double gamma_2;

    auto gammas = qr_two_iterations(*Y, y_k);
    gamma_1 = gammas.first;
    gamma_2 = gammas.second;

    auto gamma_0 = -(gamma_1 + gamma_2 + gamma_3);
    auto beta_0 = gamma_1 + gamma_2 + gamma_3;
    auto beta_1 = gamma_2 + gamma_3;
    auto beta_2 = gamma_3;

    auto b0 = beta_0 * x_2;
    auto b1 = beta_1 * x_1;
    auto b2 = beta_2 * x_k;
    unique_ptr<vec> x (new vec(b0 + b1 + b2));
    return x;
}

vec* build_uniform(int size) {
    auto value = 1.0 / size;
    return new vec(size, value);
}

// P is assumed already transposed
vec power_quad(sparse_matrix& P_t, double epsilon, int quad_frequency=5, int quad_modulo=0) {
    auto k = 1;
    shared_ptr<vec> x (build_uniform(P_t.m));
    shared_ptr<vec> x_3 = x;
    shared_ptr<vec> x_2 = x;
    shared_ptr<vec> x_1 = x;
    shared_ptr<vec> x_k = x;
    vec* v = build_uniform(x->size());

    double delta = 1000000.0;

    while(delta >= epsilon && k < MAX_ITERS) {
        unique_ptr<vec> mult (P_t.mult(*x_k));
        unique_ptr<vec> y (vec_mul(REMAIN_FACTOR, *mult));

        double w = norm(*x_k, 1) - norm(*y, 1);
        unique_ptr<vec> w_v (vec_mul(w, *v));
        sum_inplace(*y, *w_v);
        x_k = move(y);

        if (k % quad_frequency == quad_modulo) {
            cout << "Performing extrapolation..." << endl;
            x_k = quad_extrapolation(*x_3, *x_2, *x_1, *x_k);
        }

        delta = norm(*x_k - *x_1, 2);

        x_3 = x_2;
        x_2 = x_1;
        x_1 = x_k;

        cout << "Iter " << k << ", delta=" << delta << endl;
        k++;
    }

    delete v;

    return (*x_k) / norm(*x_k);
}

sparse_matrix* load_matrix(string filename) {
    ifstream file(filename);

    if (!file.is_open()) {
        cout << "No se pudo abrir el archivo " << filename << endl;
        exit(1);
    }

    double n;
    double links;

    file >> n;
    file >> links;

    cout << "Loading data..." << endl;
    auto matrix = new sparse_matrix(n, n);

    int i, j;

    auto colnum_count = vector<int>(n, 0);

    for (auto k = 0; k < links; k++) {
        file >> i;
        file >> j;

        // Matrix is built with transposed indices
        matrix->put(j - 1, i - 1, 1);
        colnum_count[i -1]++;
    }

    for (auto row: matrix->get_data()) {
        for (auto col: row.second) {
            matrix->put(row.first, col.first, 1.0 / colnum_count[col.first]);
        }
    }

    return matrix;
}

int main(int argc, char ** argv) {
    if (argc < 2) {
        cout << "Usage: tp3 [matrix_filename]" << endl;
        exit(1);
    }

    string filename = argv[1];
    unique_ptr<sparse_matrix> matrix  (load_matrix(filename));

    // cout << *matrix << endl;
    // vec* initial = build_uniform(matrix->n);
    auto solution = power_quad(*matrix, 0.000001);

    //cout << "Solution: " << solution << endl;
     solution = solution / norm(solution);
     cout << solution[0] << ", " << solution[1] << ", " << solution[2] << "...." << endl;
     cout << solution[solution.size() - 4] << ", " << solution[solution.size() - 3] << ", " << solution[solution.size() - 2] << ", " << solution[solution.size() - 1] << endl;
    return 0;
}
